spec/models/manageiq/providers/base_manager/refresher_spec.rb

```diff
@@ -0,0 +1,82 @@
+RSpec.describe ManageIQ::Providers::BaseManager::Refresher do
+  context "#initialize" do
+    let(:ems1) { FactoryBot.create(:ext_management_system) }
+    let(:ems2) { FactoryBot.create(:ext_management_system) }
+    let(:vm1)  { FactoryBot.create(:vm, :ext_management_system => ems1) }
+    let(:vm2)  { FactoryBot.create(:vm, :ext_management_system => ems2) }
+
+    it "groups targets by ems" do
+      refresher = described_class.new([vm1, vm2])
+      expect(refresher.targets_by_ems_id.keys).to include(ems1.id, ems2.id)
+    end
+  end
+
+  context "#preprocess_targets" do
+    let(:not_vm) { FactoryBot.create(:orchestration_stack) }
+    let(:ems) { FactoryBot.create(:ext_management_system) }
+    let(:vm)  { FactoryBot.create(:vm, :ext_management_system => ems) }
+    let(:lots_of_vms) do
+      num_targets = Settings.ems_refresh.full_refresh_threshold + 1
+      Array.new(num_targets) { FactoryBot.create(:vm, :ext_management_system => ems) }
+    end
+
+    it "sans sub_ems_targets" do
+      refresher = described_class.new([not_vm]) # this probably shouldn't be a stack anyway
+      refresher.preprocess_targets
+
+      expect(refresher.targets_by_ems_id[vm.ext_management_system.id]).to eq([])
```

Comment:
Not sure if this test is testing what you think, your stack doesn't have an ems_id that's why `refresher.targets_by_ems_id[vm.ext_management_system.id]` is `[]`